core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -642,9 +672,95 @@ public CompletableFuture<Optional<Throwable>> acknowledge(String memberId, List<
             lock.writeLock().unlock();
         }
 
+        if (canMoveStartOffset) {
+            maybeUpdateCachedStateAndOffsets();
+        }
+
         return CompletableFuture.completedFuture(Optional.ofNullable(throwable));
     }
 
+    // Visible for testing
+    public long startOffset() {
+        return this.startOffset;
+    }
+
+    // Visible for testing
+    public long endOffset() {
+        return this.endOffset;
+    }
+
+    public boolean canAcquireMore() {
+        lock.readLock().lock();
+        long numRecords = this.endOffset - this.startOffset + 1;
+        lock.readLock().unlock();
+        return numRecords < maxInFlightMessages;
+    }
+
+    public void maybeUpdateCachedStateAndOffsets() {
+        lock.writeLock().lock();
+        try {
+            long lastOffsetBeforeStartOffset = -1;
+            for (NavigableMap.Entry<Long, InFlightBatch> entry : cachedState.entrySet()) {
+                InFlightBatch inFlightBatch = entry.getValue();
+                if (inFlightBatch.offsetState == null) {
+                    if (inFlightBatch.inFlightState.state == RecordState.ACKNOWLEDGED ||
+                            inFlightBatch.inFlightState.state == RecordState.ARCHIVED) {
+                        lastOffsetBeforeStartOffset = inFlightBatch.lastOffset;
+                    } else {
+                        break;
+                    }
+                } else {
+                    boolean toBreak = false;
+                    for (Map.Entry<Long, InFlightState> offsetState : inFlightBatch.offsetState.entrySet()) {
+                        if (offsetState.getValue().state == RecordState.ACKNOWLEDGED ||
+                                offsetState.getValue().state == RecordState.ARCHIVED) {
+                            lastOffsetBeforeStartOffset = offsetState.getKey();
+                        } else {
+                            toBreak = true;
+                            break;
+                        }
+                    }
+                    if (toBreak) break;
+                }
+            }
+
+            // If the acknowledgement failed, then we cannot move the startOffset
```

Comment:
I didn't understand the comment. Do you mean if there is nothing to move forward?